File: Python3/94.Binary_Tree_Inorder_Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        if (root == None):
            return []
        left = self.inorderTraversal(self, root.left)
        right = self.inorderTraversal(self, root.right)
        res = left + [root.val] + right
        return res


class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        current = root
        res = []
        stack = []

        while (True):
            if current != None:
                stack.append(current)
                current = current.left
            elif (stack):
                current = stack.pop()
                res.append(current.val)
                current = current.right
            else:
                break

        return res

File: Java/94.Binary_Tree_Inorder_Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        if (root == null) {
            return new List<Integer>();
        }
        List<Integer> left = inorderTraversal(root.left);
        List<Integer> right = inorderTraversal(root.right);
        List<Integer> res = new List<>();
        res.addAll(left);
        res.add(root.val);
        res.addAll(right);
        return res;
    }
}

class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        TreeNode current = root;
        Stack<TreeNode> stack = new Stack<>();
        List<Integer> res = new ArrayList<>();

        while (true) {
            if (current != null) {
                stack.push(current);
                current = current.left;
            } else if (!stack.isEmpty()) {
                current = stack.pop();
                res.add(current.val);
                current = current.right;
            } else {
                break;
            }
        }
        
        return res;
    }
}